File: cpp_engine/src/converter.cpp
#include "converter.h"
#include "config.h"
#include "security.h"
#include "formats.h"
#include "gpu.h"
#include "media.h"
#include <sstream>

namespace engine {

const std::unordered_map<std::string, std::string> ConverterWorker::CODEC_MAP = {
    {"libx264", "h264"}, {"h264_nvenc", "h264"}, {"h264_amf", "h264"}, {"h264_qsv", "h264"},
    {"libx265", "hevc"}, {"hevc_nvenc", "hevc"}, {"hevc_amf", "hevc"}, {"hevc_qsv", "hevc"},
    {"libvpx-vp9", "vp9"}, {"av1_nvenc", "av1"}, {"av1_amf", "av1"}, {"av1_qsv", "av1"},
};

ConverterWorker::ConverterWorker(const std::string& input_path, const std::string& output_path,
                                 const std::string& output_format, bool dev_mode,
                                 bool use_gpu, const std::string& preferred_encoder)
    : output_format_(output_format), dev_mode_(dev_mode),
      use_gpu_(use_gpu), preferred_encoder_(preferred_encoder) {
    this->input_path = validate_path(input_path, "input_path");
    this->output_path = validate_output_path(output_path, "output_path");
    validate_string(output_format_, "output_format", 32);
    init_process();
}

std::optional<std::string> ConverterWorker::codec_for_encoder(const std::string& encoder) const {
    auto it = CODEC_MAP.find(encoder);
    return it != CODEC_MAP.end() ? std::make_optional(it->second) : std::nullopt;
}

std::vector<std::string> ConverterWorker::build_command() {
    auto file_type = detect_file_type(input_path);
    auto& fmt = output_format_;
    std::vector<std::string> output_args = {"-map_metadata", "0"};
    bool cuda = false;

    auto& audio_formats = audio_output_formats();
    auto& photo_formats = photo_output_formats();
    auto& video_formats = video_output_formats();

    auto it_audio = audio_formats.find(fmt);
    auto it_photo = photo_formats.find(fmt);
    auto it_video = video_formats.find(fmt);

    if (it_audio != audio_formats.end() && (file_type == "video" || file_type == "audio" || dev_mode_)) {
        auto& options = it_audio->second;
        output_args.insert(output_args.end(), {"-map", "0:a:0", "-vn"});
        if (options.acodec) output_args.insert(output_args.end(), {"-c:a", *options.acodec});
        if (options.bitrate) output_args.insert(output_args.end(), {"-b:a", *options.bitrate});

    } else if (it_photo != photo_formats.end() && (file_type == "photo" || (dev_mode_ && !video_formats.count(fmt)))) {
        auto& options = it_photo->second;
        output_args.insert(output_args.end(), {"-frames:v", "1"});
        if (options.quality) output_args.insert(output_args.end(), {"-q:v", *options.quality});
        if (options.compression) output_args.insert(output_args.end(), {"-compression_level", *options.compression});

    } else if (it_video != video_formats.end() && (file_type == "video" || file_type == "photo" || dev_mode_)) {
        auto& options = it_video->second;
        auto encoder = get_video_encoder(use_gpu_, options.vcodec.value_or("libx264"),
                                         preferred_encoder_, fmt);
        MediaMetadata metadata;
        if (file_type == "video") {
            metadata = probe_media(input_path);
        }

        bool copy_video = (file_type == "video" && !max_width_ &&
                           metadata.vcodec.has_value() &&
                           metadata.vcodec == codec_for_encoder(encoder));

        output_args.insert(output_args.end(), {"-map", "0:V:0", "-map", "0:a:0?"});

        if (copy_video) {
            output_args.insert(output_args.end(), {"-c:v", "copy"});
        } else {
            cuda = uses_cuda_frames(encoder, metadata);
            if (max_width_) {
                output_args.push_back("-vf");
                output_args.push_back(scale_filter(*max_width_, -2, cuda, true));
            }
            auto enc_args = video_encoding_args(encoder, 18);
            output_args.insert(output_args.end(), enc_args.begin(), enc_args.end());
        }

        if (options.acodec) {
            std::string audio_codec = *options.acodec;
            if (audio_codec == "libopus") audio_codec = "opus";
            else if (audio_codec == "libmp3lame") audio_codec = "mp3";

            if (copy_video && metadata.acodec && *metadata.acodec == audio_codec) {
                output_args.insert(output_args.end(), {"-c:a", "copy"});
            } else {
                output_args.insert(output_args.end(), {"-c:a", *options.acodec, "-b:a", "192k"});
            }
        } else {
            output_args.push_back("-an");
        }
    } else {
        throw std::runtime_error("Unsupported file type '" + file_type + "' for conversion.");
    }

    if (fmt == "mp4" || fmt == "mov" || fmt == "m4v") {
        output_args.insert(output_args.end(), {"-movflags", "+use_metadata_tags"});
    }

    auto cuda_args = decode_args(cuda);
    std::vector<std::string> cmd = {find_binary("ffmpeg"), "-y", "-nostdin", "-hide_banner"};
    cmd.insert(cmd.end(), cuda_args.begin(), cuda_args.end());
    cmd.push_back("-i");
    cmd.push_back(input_path);
    cmd.insert(cmd.end(), output_args.begin(), output_args.end());
    cmd.push_back(output_path);
    return cmd;
}

}

File: PyEngine/workers/converter.py
"""File format conversion using ffmpeg."""
from PyEngine.core.config import find_binary
from PyEngine.core.security import validate_path, validate_output_path, validate_string
from typing import List, Optional

from PyEngine.workers.ffmpeg import FfmpegWorker
from PyEngine.core.media import probe_media, uses_cuda_frames, decode_args, scale_filter


def find_ffmpeg() -> str:
    return find_binary("ffmpeg")


class ConverterWorker(FfmpegWorker):
    def __init__(
        self,
        input_path: str,
        output_path: str,
        output_format: str,
        dev_mode: bool = False,
        use_gpu: bool = False,
        preferred_encoder: str = "",
    ):
        self.input_path = validate_path(input_path, "input_path")
        self.output_path = validate_output_path(output_path, "output_path")
        self.output_format = validate_string(output_format, "output_format", 32)
        self.dev_mode = dev_mode
        self.use_gpu = use_gpu
        self.preferred_encoder = preferred_encoder
        self.max_width: Optional[int] = None
        self._init_process()

    _CODEC_MAP = {
        'libx264': 'h264', 'h264_nvenc': 'h264', 'h264_amf': 'h264', 'h264_qsv': 'h264',
        'libx265': 'hevc', 'hevc_nvenc': 'hevc', 'hevc_amf': 'hevc', 'hevc_qsv': 'hevc',
        'libvpx-vp9': 'vp9', 'av1_nvenc': 'av1', 'av1_amf': 'av1', 'av1_qsv': 'av1',
    }

    def _codec_for_encoder(self, encoder: str) -> Optional[str]:
        return self._CODEC_MAP.get(encoder)

    def _probe_codecs(self) -> Optional[dict]:
        return probe_media(self.input_path) or None

    def _build_command(self) -> List[str]:
        from PyEngine.formats.video import VIDEO_OUTPUT_FORMATS
        from PyEngine.formats.photo import PHOTO_OUTPUT_FORMATS
        from PyEngine.formats.audio import AUDIO_OUTPUT_FORMATS
        from PyEngine.formats.detection import detect_file_type
        from PyEngine.core.gpu import get_video_encoder, video_encoding_args

        file_type = detect_file_type(self.input_path)
        fmt = self.output_format
        output_args = ['-map_metadata', '0']
        cuda = False
        if fmt in AUDIO_OUTPUT_FORMATS and (file_type in ('video', 'audio') or self.dev_mode):
            options = AUDIO_OUTPUT_FORMATS[fmt]
            output_args += ['-map', '0:a:0', '-vn']
            if options.get('acodec'):
                output_args += ['-c:a', options['acodec']]
            if options.get('bitrate'):
                output_args += ['-b:a', options['bitrate']]
        elif fmt in PHOTO_OUTPUT_FORMATS and (file_type == 'photo' or (self.dev_mode and fmt not in VIDEO_OUTPUT_FORMATS)):
            options = PHOTO_OUTPUT_FORMATS[fmt]
            output_args += ['-frames:v', '1']
            if 'quality' in options:
                output_args += ['-q:v', options['quality']]
            if 'compression' in options:
                output_args += ['-compression_level', options['compression']]
        elif fmt in VIDEO_OUTPUT_FORMATS and (file_type in ('video', 'photo') or self.dev_mode):
            options = VIDEO_OUTPUT_FORMATS[fmt]
            encoder = get_video_encoder(self.use_gpu, fallback=options['vcodec'],
                                        preferred_encoder=self.preferred_encoder, output_format=fmt)
            metadata = (self._probe_codecs() or {}) if file_type == 'video' else {}
            copy_video = (file_type == 'video' and not self.max_width
                          and metadata.get('vcodec') is not None
                          and metadata.get('vcodec') == self._codec_for_encoder(encoder))
            output_args += ['-map', '0:V:0', '-map', '0:a:0?']
            if copy_video:
                output_args += ['-c:v', 'copy']
            else:
                cuda = uses_cuda_frames(encoder, metadata)
                if self.max_width:
                    output_args += ['-vf', scale_filter(self.max_width, cuda=cuda, limit=True)]
                output_args += video_encoding_args(encoder, quality=18)
            audio_encoder = options.get('acodec')
            if audio_encoder:
                audio_codec = {'libopus': 'opus', 'libmp3lame': 'mp3'}.get(audio_encoder, audio_encoder)
                # Copy only the format's chosen audio codec, never an arbitrary
                # codec from a source container with broader codec support.
                if copy_video and metadata.get('acodec') == audio_codec:
                    output_args += ['-c:a', 'copy']
                else:
                    output_args += ['-c:a', audio_encoder, '-b:a', '192k']
            else:
                output_args += ['-an']
        else:
            raise ValueError(f"Unsupported file type '{file_type}' for conversion. Please select a supported output format.")

        if fmt in ('mp4', 'mov', 'm4v'):
            output_args += ['-movflags', '+use_metadata_tags']
        return [find_ffmpeg(), '-y', '-nostdin', '-hide_banner', *decode_args(cuda),
                '-i', self.input_path, *output_args, self.output_path]
